features/events/api/src/main/java/org/opennms/netmgt/xml/event/ImmutableAlarmData.java

```diff
@@ -0,0 +1,266 @@
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2020 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2020 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.xml.event;
+
+import org.opennms.core.utils.ImmutableCollections;
+import org.opennms.core.utils.MutableCollections;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * An immutable implementation of '{@link IAlarmData}'.
+ */
+public final class ImmutableAlarmData implements IAlarmData {
+    private final String reductionKey;
+    private final Integer alarmType;
+    private final String clearKey;
+    private final Boolean autoClean;
+    private final String x733AlarmType;
+    private final Integer x733ProbableCause;
+    private final List<IUpdateField> updateFieldList;
+    private final IManagedObject managedObject;
+
+    private ImmutableAlarmData(Builder builder) {
+        reductionKey = builder.reductionKey;
+        alarmType = builder.alarmType;
+        clearKey = builder.clearKey;
+        autoClean = builder.autoClean;
+        x733AlarmType = builder.x733AlarmType;
+        x733ProbableCause = builder.x733ProbableCause;
+        updateFieldList = ImmutableCollections.with(ImmutableUpdateField::immutableCopyFrom)
+                .newList(builder.updateFieldList);
+        managedObject = ImmutableManagedObject.immutableCopy(builder.managedObject);
+    }
+
+    public static Builder newBuilder() {
+        return new Builder();
+    }
+
+    public static Builder newBuilderFrom(IAlarmData alarmData) {
+        return new Builder(alarmData);
+    }
+
+    public static IAlarmData immutableCopy(IAlarmData alarmData) {
+        if (alarmData == null || alarmData instanceof ImmutableAlarmData) {
+            return alarmData;
+        }
+        return newBuilderFrom(alarmData).build();
+    }
+
+    public static IAlarmData fromMutable(AlarmData alarmData) {
```

Comment:
Looking at this, I think there is value in updating the original bean implementations to implement the new interface. That way these `fromMutable()` methods can be replaced with usage of `newBuilderFrom().build()`.

This applies to all the new immutables.

Reply:
Also this removes the dependency of the jaxb type from the immutable types in case we want to separate them.

Reply:
Gave this a try - but unfortunately didn't work out.  Will go with your earlier suggestion.